src/config.rs

```diff
@@ -0,0 +1,193 @@
+                |UserPrincipal {
+                     name,
+                     password,
+                     kvno,
+                 }| {
+                    let salt = format!("{realm}{name}");
+
+                    let base_key =
+                        DerivedKey::new_aes256_cts_hmac_sha1_96(password, &salt, None, *kvno)?;
+
+                    let princ_name = Name::principal(name, realm);
+
+                    Ok((
+                        princ_name,
+                        PrincipalRecord {
+                            service: false,
+                            base_key,
+                        },
+                    ))
+                },
+            )
+            .chain(service.iter().map(
+                |ServicePrincipal {
+                     srvname,
+                     hostname,
+                     password,
+                     kvno,
+                 }| {
+                    let name = format!("{srvname}/{hostname}");
+                    let salt = format!("{realm}{name}");
+
+                    let princ_name = Name::service(srvname, hostname, realm);
+
+                    let base_key =
+                        DerivedKey::new_aes256_cts_hmac_sha1_96(password, &salt, None, *kvno)?;
+
+                    Ok((
+                        princ_name,
+                        PrincipalRecord {
+                            service: true,
+                            base_key,
+                        },
+                    ))
+                },
+            ))
+            .collect::<Result<_, KrbError>>()?;
+
+        let allowed_clock_skew = Duration::from_secs(300);
+
+        // Short tickets are good, but not too short.
+        let ticket_min_lifetime = Duration::from_secs(60);
+        // Should be short-ish to promote frequent renewals.
+        let ticket_granting_ticket_lifetime = Duration::from_secs(900);
+        // You can renew for up to a week. We may change this ....
+        let ticket_max_renew_time = Some(Duration::from_secs(86400 * 7));
+
+        // This needs to be *long* because TGS can't be renewed, and most
+        // applications absolutely *lose their mind* when this expires rapidly.
+        //
+        // For example, finder on macos with SMB just disconnects you soon after this
+        // expires, doesn't try to get a new TGS and open a new session to continue
+        // the connection or make a new connection.
+        //
+        // The only way to make this work is for the TGS to be looooooooooongggggggg.
+        // I'm really not sure what value should be because if it's too long then that
+        // allows an attacker a long-term access to the service. If it's too short,
+        // then things break.
+        //
+        // Likely we need to ensure that there is a way to kill sessions on the SMB side
+        // when we expire an account. Thought needed to what this value should be ...
+        let service_granting_ticket_lifetime = Duration::from_secs(3600 * 8);
```

Comment:
May be a stupid question, but what does AD (or Samba AD) do here? I assume they would have the same problem.

Reply:
@2xsaiko In my testing with short TGS lifetimes, you get a lot of weird problems, because while a TGT can "renew", a TGS can't so it disconnects and reconnects the session for SMB "at best". Otherwise it can leave you hanging or it just hard disconnects you with no reconnect. It's why we had to extend this to be longer. I would love it to be shorter (similar to an oauth2 access token) since a TGS once issued can't be revoked.

Reply:
So they also just have long lifetimes for these? That sucks, but alright.